20220714-lookup-api.md

```diff
@@ -0,0 +1,457 @@
+
+- **What use cases does it support?**
+  Search with Permissions where the number of objects of a particular type the user could have access to is low (~ thousands), and the percentage of the total objects that the user can have access to is also low.
+
+- **What is the expected outcome?**
+  OpenFGA clients use the Lookup API to return a list of object ids and then the client uses the list to filter and order their object list in their applications (pagination on the filtered list is the clients responsibility). The expectation is that the client consumes the whole Lookup response before filtering client side, and does not try to partially filter based on a subset of the response from Lookup. Hence, it is not the goal of Lookup to be able to continue the Lookup process from a prior invocation.
+
+# What it is
+[what-it-is]: #what-it-is
+
+The Lookup API will provide clients with the ability to query the object ids that a user has a particular relationship with. The list of object ids the Lookup API returns can then be used by a client to filter an object collection to only the objects that a user has a particular relationship with by taking the intersection of it with any client provided filters/ordering of the underlying dataset. This is described in more detail with [Search with Permission (Option 3)](https://openfga.dev/docs/interacting/search-with-permissions#option-3-build-a-list-of-ids-then-search) document.
+
+It will be provided as two different API endpoints: one will provide a streaming endpoint and the other will implement the endpoint without streaming. These two different endpoints could be used interchangably, but clients can choose which one may be better for their integration environment.
+
+Consider a developer/client integrating with OpenFGA with the following authorization model:
+```
+type folder
+    relations
+        define viewer as self
+
+type document
+    relations
+        define viewer as self or editor or viewer from parent
+        define editor as self
+        define parent as self
+```
+and the following relationship tuples:
+```
+document:doc1, viewer, bob
+document:doc2, editor, bob
+document:doc3, parent, folder:folder1
+folder:folder1, viewer, bob
+```
+
+## StreamedLookup
+StreamedLookup will implement a grpc [server streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc) endpoint.
+
+The endpoint will stream each object id the target user has the given relationship with as it is evaluated, and the server will continuing streaming results until either the lookup has been exhausted or the `streamed-lookup-deadline` has been reached (see [Configuration Changes](#configuration-changes)).
+
+Sample grpcurl command:
+```
+grpcurl -plaintext \
+  -d '{
+           "store_id":"<store>", 
+           "authorization_model_id":"<modelID>",
+           "object_type":"document", 
```

Comment:
To be consistent with our other endpoints, can we wrap the tuple inside a `tuple_key` object?

E.g. take a look at the [Read endpoint](https://openfga.dev/api/service#/Relationship%20Tuples/Read)

```js
{
           "store_id":"<store>", 
           "authorization_model_id":"<modelID>",
           "tuple_key": {
              "user": ...
              "relation": ...
              "object_type": ...
           }
}
```

Reply:
Also we do not use `object_type` anywhere else in the exposed interface, we use `type` (e.g. in the authorization model) or "object": "$TYPE:" (e.g. in Read)

Reply:
Using a `tuple_key` here doesn't make sense to me, because we're not querying a specific tuple key (e.g. object, relation, user). We're querying all tuples matching a filter where user and relation and object type is provided.

I'm curious what other people think about this?

Reply:
For the purposes of representing properties (because not all 3 fields might be fully defined) of `tuple` OpenFGA uses the `tuple_key` proto. Same logic as `READ`, we are looking for subset of the composite entity. No need to include another entity when the existing one works. 

> Also we do not use object_type anywhere else in the exposed interface, we use type (e.g. in the authorization model) or "object": "$TYPE:" (e.g. in Read)

Agreed, OpenFGA uses `type`.

Reply:
Tho will admit that having to pass:
`"type": "document:"` (namely the `:`) is not the greatest DX.

Reply:
> Tho having to pass: `"type": "document:"` (namely the `:`) is not the greatest DX.

That's the primary rationale for my opinion on this. The DX of our tuple key is obscure at best. If I want to query a subset of the composite entity, I have to pass a string like "document:", which isn't clear at all. I think separating them into their individual parts makes more sense from a query perspective.

Reply:
IMO, from an end user perspective, the lookup and read's parameter should be very similar as they deal with "looking up" relations for a particular type.  If we are to have lookup with a new syntax call object_type, we should have similar thing for read.  Otherwise, it feels there are inconsistency in the API. Further, if we are to introduce `object_type`, we should group all three (object_type, relation, user) into a high entity.  This will allow us to have multiples of these groupings down the road.

Reply:
@adriantam the semantics of Lookup are different than Read though, so I think comparing them is not the best comparison. For example, the Read API allows you to read all tuples of a particular object type or read specific tuples of a particular (singular) object.

Lookup looks up all of the objects of only a particular type. You don't lookup a specific object (otherwise that would be a Check). 

@rhamzeh @elbuo8 - should we just go with `type` for now so that it's more consistent with our existing API but it doesn't introduce the DX shortcomings of the composite `tuple_key` approach? WDYT?

Reply:
`type` 👍🏽

Reply:
I'd prefer to not to use the `tuple_key` and use `type`

Reply:
Sounds good. I've made this change to the API structure. Marking as resolved for now.